---
title: "**P1** | Basic tools for data visualization"
subtitle: "Introduction"
author: "<b>Marta Coronado Zamora and Adrià Auladell</b>"
date: "20 September 2024"
output:
  xaringan::moon_reader:
    css: ["default", "rutgers", "metropolis-fonts", "https://cdnjs.cloudflare.com/ajax/libs/animate.css/3.7.2/animate.min.css", "https://use.fontawesome.com/releases/v5.10.1/css/all.css"]
    lib_dir: libs
    self_contained: yes
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
      ratio: "16:9"
      
---
```{r setup, include=FALSE}
options(htmltools.dir.version = FALSE)
# Animations: https://github.com/daneden/animate.css#animations
knitr::opts_chunk$set(echo = FALSE, eval = TRUE, fig.width = 4, fig.height = 3.5, fig.show = 'hold', message = FALSE)
library(ggplot2)
library(lattice)
library(tidyverse)
# pagedown::chrome_print("P1-slides.html")
```

class: center, animated, bounceInDown

<style>
.title-slide {
  background-image: url(img/logoesci-fons-blanc.png);
  background-size: 150px;
}
</style>

# Keep in touch

| Adrià Auladell |
|:-:|
| <a href="mailto:adria.auladell@ibe.upf-csic.es"><i class="fa fa-paper-plane fa-fw"></i> adria.auladell@ibe.upf-csic.es </a> | 
|  <a href="https://www.ibe.upf-csic.es/"><i class="fa fa-map-marker fa-fw"></i>&nbsp; Institut de Biologia Evolutiva (UPF-CSIC) </a> |

---
layout: true
class: animated, fadeIn

---
# Practical session dynamics

**Content** (P1-P5)
- Introduction 
- Exercises - complete and submit to [aul@-ESCI](https://aula.esci.upf.edu/course/view.php?id=8592)
- Project (divided in to 2 assignments)
<br><br>

--

**Interactive <i class="fab fa-r-project fa-fw"></i></i> documents**  
`R code` can be executed within RStudio!

```{r, echo=TRUE}
value <- 2
value + 3
```

---
layout: false
class: left, bottom, inverse, animated, bounceInDown

# Get started!
## **Tools for data visualization**

---
layout: true
class: animated, fadeIn

---

## Type of tools

### Two main types:

- Graphical user interface (GUI)
  <font color="#A8A8A8">Many examples: Perseus computational platform, Cytoscape, Blast2GO, Gephi, ... </font>
  
- Code-based
  <font color="#A8A8A8">R (and other computer languages)</font>

[Wide range...](https://en.wikipedia.org/wiki/List_of_information_graphics_software)

<br>
--

<div style="background-color:#FFDA9E">
<b><i class="fas fa-question-circle"></i>&nbsp;Question</b>
<br>
What prons and cons do you think GUI tools have in comparison to code-based? 
</div>

---

## Visualization libraries in R

- `base`
- `grid`: `lattice` and `ggplot2`

<center>
```{r, warning = FALSE, echo = FALSE}

# Basic R boxplot
boxplot(Petal.Length ~ Species,
        data = iris,
        ylab = "Petal length",
        col = c("red", "blue", "green"),
        main = "Basic")

# Lattice boxplot
bwplot(
  Petal.Length ~ Species,
  group = Species,
  data = iris,
  main = "Lattice",
  ylab = "Petal length",
  par.settings = list(box.umbrella=list(col = c("red", "green", "blue")),
                      box.dot = list(col = c("red", "green", "blue")),
                      box.rectangle = list(col = c("red", "green", "blue"))))

# ggplot2 boxplot
ggplot(data = iris, mapping = aes(x = Species, y = Petal.Length, fill = Species)) +
  geom_boxplot() +
  ylab("Petal length") +
  ggtitle("ggplot2")
```

<br>
--

<div align="left" style="background-color:#FFDA9E">
<b><i class="fas fa-question-circle"></i>&nbsp;Question</b>
<br>
Describe the graphics. In your opinion, which do you think is the simplest? and the most complex? do you think the code to generate the figures reflect the complexity?
</div>

---
## Visualization libraries in R

```{r, echo=TRUE, message=FALSE, eval = FALSE}
# base
hist(iris$Sepal.Width)

# ggplot2
ggplot(iris, aes(Sepal.Width)) +
  geom_histogram()
```

<center>
```{r, echo=FALSE, message=FALSE}
# base
hist(iris$Sepal.Width)

# ggplot2
ggplot(iris, aes(Sepal.Width)) +
  geom_histogram()
```

---
## Visualization libraries in R

.pull-left[
```{r, echo = TRUE, eval = FALSE}
# base
plot(circumference ~ age,
     data=Orange[Orange$Tree %in% "4", ], type = "l",
     main = "Base - complex")
points(circumference ~ age, col="darkred",
       data=Orange[Orange$Tree %in% "2", ], type = "l")
points(circumference ~ age, col="orange",
       data=Orange[Orange$Tree %in% "5", ], type = "l")
points(circumference ~ age, col="yellow",
       data=Orange[Orange$Tree %in% "1", ], type = "l")
points(circumference ~ age, col="darkgreen",
       data=Orange[Orange$Tree %in% "3", ], type = "l")
legend("topleft",
       c("4", "2", "5", "1", "3"), title="Tree",
       col=c("black", "darkred", "darkorange", "yellow", "darkgreen"),
       lty=c(1, 1, 1, 1, 1))
```

```{r, echo = TRUE, eval = FALSE}
# ggplot2
ggplot(Orange, aes(age, circumference,
      colour = Tree)) + geom_line() +
      labs(title = "ggplot2 - complex")
```
]

--
.pull-right[
```{r, echo=FALSE, fig.keep='last', fig.width=5, fig.height=3.5, fig.align="center"}
# base
plot(circumference ~ age,
     data=Orange[Orange$Tree %in% "4", ], type = "l",
     main = "Base - complex")
points(circumference ~ age, col="darkred",
       data=Orange[Orange$Tree %in% "2", ], type = "l")
points(circumference ~ age, col="orange",
       data=Orange[Orange$Tree %in% "5", ], type = "l")
points(circumference ~ age, col="yellow",
       data=Orange[Orange$Tree %in% "1", ], type = "l")
points(circumference ~ age, col="darkgreen",
       data=Orange[Orange$Tree %in% "3", ], type = "l")
legend("topleft",
       c("4", "2", "5", "1", "3"), title="Tree",
       col=c("black", "darkred", "darkorange", "yellow", "darkgreen"),
       lty=c(1, 1, 1, 1, 1))
```

```{r, echo=FALSE, fig.width=5, fig.height=2.5, fig.align="center"}
# ggplot2
ggplot(Orange, aes(age, circumference, colour = Tree)) +
  geom_line()+
  labs(title = "ggplot2 - complex")
```
]


---
### Other visualization libraries

(Outside our scope)

- Python
    + `matplotlib`, `seaborn`
    + `Bokeh`, `pygal`
- Java: `Processing`
- Javascript: `D3.js`

<center>
<img src="img/bokeh.png" alt="Bokeh plots" width="500"/>
</center>

---
layout: false
class: left, bottom, inverse, animated, bounceInDown

# Basic R knowledge

---
layout: true
class: animated, fadeIn

---
### Installing a package

```{r, eval = FALSE, echo = TRUE}
# Download and install a package from CRAN
install.packages("ggplot2")

# Download and install a package from GitHub(you need the devtools library installed)
devtools::install_github("yihui/xaringan")
```

### Loading a package

```{r, eval = FALSE, echo = TRUE}
# Load the library to the current session
library("ggplot2")
library("xaringan")
```

### Loading data

```{r, eval = FALSE, echo = TRUE}
# Loading a tab-separated file with a header
data <- read.table("data.txt", header = TRUE, sep = "\t")
```

---
## Data types and structures

Main data types <font color=" #A8A8A8">(other will not be discussed: _complex_ and _raw_)</font>:

- __Logical__: can only take on two values: true (`TRUE`, `T`) or false (`FALSE`, `F`)
- __Numeric__: real or decimal (`2`, `15.5`)
- __Integer__: `2L` (the `L` tells R to store this as an integer)
- __Character__: any type of character or number (`"a"`, `"swc"`, `"2"`)

--

<i class="fas fa-info-circle"></i> To know the data type, you can use the `class()` function.


```{r, echo = TRUE}
type_list <- list(TRUE, 1.2, 10L, "a")
sapply(type_list, class)
```

---
## Data types and structures

Elements of the previous data types may be combined to form data structures. Main structures:

- __Vector__: collection of elements that holds data of a single data type
- __Matrix__: vector with dimensions (the number of rows and columns)
- __Factor__: to deal with categorical variables
- __List__: a special type of vector where each element can be a different type
- __Data Frame__ <i class="fas fa-star"></i>: a special type of list where every element of the list has same length 

--
```{r, echo = TRUE}
# A vector x of mode numeric
x <- c(1, 2, 3)

# A vector y of mode logical
y <- c(TRUE, TRUE, FALSE, FALSE)

# A vector z of mode character
z <- c("Sarah", "Tracy", "Jon")
```

---
## Data types and structures

Elements of the previous data types may be combined to form data structures. Main structures:

- __Vector__: collection of elements that holds data of a single data type
- __Matrix__: vector with dimensions (the number of rows and columns)
- __Factor__: to deal with categorical variables
- __List__: a special type of vector where each element can be a different type
- __Data Frame__ <i class="fas fa-star"></i>: a special type of list where every element of the list has same length 

```{r, echo = TRUE}
matrix22 <- matrix(
  c(1, 2, 3, 4),
  nrow = 2,
  ncol = 2)
matrix22
```

---
## Data types and structures

Elements of the previous data types may be combined to form data structures. Main structures:

- __Vector__: collection of elements that holds data of a single data type
- __Matrix__: vector with dimensions (the number of rows and columns)
- __Factor__: to deal with categorical variables
- __List__: a special type of vector where each element can be a different type
- __Data Frame__ <i class="fas fa-star"></i>: a special type of list where every element of the list has same length 

```{r, echo = TRUE}
factor_vector <- as.factor(c("rna", "dna", "dna", "rna"))
factor_vector
str(factor_vector)
```

---
## Data types and structures

Elements of the previous data types may be combined to form data structures. Main structures:

- __Vector__: collection of elements that holds data of a single data type
- __Matrix__: vector with dimensions (the number of rows and columns)
- __Factor__: to deal with categorical variables
- __List__: a special type of vector where each element can be a different type
- __Data Frame__ <i class="fas fa-star"></i>: a special type of list where every element of the list has same length 

```{r, echo = TRUE}
x <- list(1, "a", TRUE, 1+4i)
x
```

---
## Data types and structures

Elements of the previous data types may be combined to form data structures. Main structures:

- __Vector__: collection of elements that holds data of a single data type
- __Matrix__: vector with dimensions (the number of rows and columns)
- __Factor__: to deal with categorical variables
- __List__: a special type of vector where each element can be a different type
- __Data Frame__ <i class="fas fa-star"></i>: a special type of list where every element of the list has same length 

```{r, echo = TRUE}
dat <- data.frame(id = letters[1:10], x = 1:10, y = 11:20)
dat
```
---


## Tidy data

Data frames with one observation per row and one variable per column.

```{r, eval = TRUE}
not_tidy <- data.frame(
    student = c( "Marta", "Joan" ),
    course_age = c("1_18", "1_19"),
    First_exam = c( 9, 7 ),
    Second_exam = c( 6, 8 )
)
tidy <- data.frame(
    student = rep( c( "Marta", "Joan" ), times = 2 ),
    course = rep( c( "1", "1" ), times = 2 ),
    age = rep( c( "18", "19" ), times = 2 ),
    exam = rep( c( "First_exam", "Second_exam" ), each = 2 ),
    score = c( 9, 7, 6, 8 )
)
```

```{r, eval = TRUE, echo = TRUE}
not_tidy
tidy
```

---


## Tidy data

Data frames with one observation per row and one variable per column.

<center>
<img src="https://d33wubrfki0l68.cloudfront.net/6f1ddb544fc5c69a2478e444ab8112fb0eea23f8/91adc/images/tidy-1.png" alt="Bokeh plots" width="900"/>
</center>

---


## Tidy data

Two types of tidy data structures:

1. **Wide format** (most common): in a wide form, the multiple measures of a single observation are stored in a single row. 


```{r, eval = TRUE, echo = F}
USArrests <- datasets::USArrests %>% 
  tibble::rownames_to_column("Pop")

head(USArrests)
```


2. **Long format**: each row corresponds to one measure on one observation.


```{r, eval = TRUE, echo = F}
head(tidyr::pivot_longer(USArrests,
                    names_to = "Measure", 
                     values_to = "Value",
                     cols = -Pop) )
```


---
## Tidy data

The function to change from wide to long format has evolved to facilitate its usage: 

```{r eval = FALSE, echo = TRUE}

#Before (~circa 2015)
reshape2::melt(
    USArrests,
    id.vars= "Pop",
    variable.name = "Measure",
    value.name = "Value"
    )

# Inbetween: tidyr::gather

#After (~circa 2023)
tidyr::pivot_longer(
  USArrests, 
  names_to = "Measure",
  values_to = "Value",
  cols = -Pop
  )

```

---
# Getting help <i class="fas fa-question-circle"></i>

- `?read.table`, `?str`, `?as.factor`
- Press F1 (in RStudio)
- [Stack Overflow](https://stackoverflow.com) ([`R`](https://stackoverflow.com/questions/tagged/r), [`ggplot2`](https://stackoverflow.com/questions/tagged/ggplot2))
- Ask your classmates or your teacher

---

# Exercise: describe a data set

Read the file in this [link](https://goo.gl/c9szmq), ensure it has a tidy and long format and indicate the data type of each variable.

---

# Practice <i class="fas fa-cogs"></i>
## Introduction to `ggplot2`

- Open the document `P1_exercises.Rmd` in RStudio and complete the exercises.
- Upload the completed document to [Aul@-ESCI](https://aula.esci.upf.edu/course/view.php?id=8592) at the end of the session.

---

# Project

## Group project

The project has 3 different parts (A, B and C) divided in two big assignments.
- You can deliver the parts separately to get feedback before submitting the final version
- Each part must be submitted before next practical session
- The first assignment will contain parts A and B
- The second assignment will contain part C
- ~15 minutes in the end of each class devoted to discuss your problems

---
# Project

## Group project

__Part A__

- __1\.__ Create groups of \~4 people
- __2\.__ Choose a data set with the following requirements
    + Tabular format (txt, csv, tsv...)
    + More than 80 observations
    + At least 6 variables
    + At least 2 discrete and 3 continuous variables
    + Data with biological meaning
    + Different from the ones chosen by other groups

---
# Project

## Group project

- __3\.__ Describe your data set:
    + Where and why was the information collected?
    + Which is the meaning of each variable?
    + Do the variables have unit? Which one?
    + Does the data set have a long format?
- __4\.__ Write the code to:
    + Read it into R
    + Reshape the data if necessary into long format
    + Check the variable classes and update them if necessary

Write 3 and 4 in an `R Markdown` document and __submit it before next practical session__ (one per group).

<i class="fas fa-info-circle"></i> If you need help formatting the R Markdown, ask me for a guide of an introduction to R Markdown.

---

## Data sets from research articles

- ["Zika virus evolution and spread in the Americas"](https://www.nature.com/articles/nature22402#supplementary-information) (Table S2)
- ["Great ape genetic diversity and population history"](https://www.nature.com/articles/nature12228#supplementary-information) (Table S1 or S3)
- ["Transcriptome and genome sequencing uncovers functional variation in humans"](https://www.nature.com/articles/nature12531). [Table with cis eQTLs in EUR](https://www.ebi.ac.uk/arrayexpress/files/E-GEUV-1/EUR373.gene.cis.FDR5.best.rs137.txt.gz) ([description](https://www.ebi.ac.uk/arrayexpress/files/E-GEUV-1/GeuvadisRNASeqAnalysisFiles_README.txt))
- ["Signatures of archaic adaptive introgression in present-day human Populations"](https://academic.oup.com/mbe/article/34/2/296/2633371#supplementary-data) (Table S3)
- ["The evolutionary history of dogs in the Americas"](http://science.sciencemag.org/content/361/6397/81) (Table S1)
- ["Ancient genomes document multiple waves of migration in Southeast Asian prehistory"](http://science.sciencemag.org/content/361/6397/92) (Table S1)
- ["Population-scale long-read sequencing uncovers transposable elements associated with gene expression variation and adaptive signatures in _Drosophila_"](https://www.nature.com/articles/s41467-022-29518-8#Sec29) (Table S10)
- ["Comprehensive characterization of 536 patient-derived xenograft models prioritizes candidates for targeted treatment"](https://www.nature.com/articles/s41467-021-25177-3#Sec31) (Table S1)
- ["Pan-cancer analysis of whole genomes"](https://www.nature.com/articles/s41586-020-1969-6#Sec23) (Table S1)
- ["The genomic basis of copper tolerance in _Drosophila_ is shaped by a complex interplay of regulatory and environmental factors"](https://static-content.springer.com/esm/art%3A10.1186%2Fs12915-022-01479-w/MediaObjects/12915_2022_1479_MOESM4_ESM.xlsx) (Table S3)
- ["Transposons contribute to the diversification of the head, gut, and ovary transcriptomes across _Drosophila_ natural strains"](https://raw.githubusercontent.com/GonzalezLab/chimerics-transcripts-dmelanogaster/main/01.%20Chimeric%20gene-TE%20transcripts/chimerics_v2.tab) (Chimeric gene-TE transcripts data)

